### ISSUE TITLE
Creating COSI user causes the object store reconcile to fail several times before finally succeeding

### ISSUE THREAD

**Issue by travisn** [rook/rook#13904]
**Is this a bug report or feature request?**
* Bug Report

**Deviation from expected behavior:**
The COSI user is created with each object store creation. After the object store creation is completed, the controller attempts to create the COSI user and fails the reconcile. After repeated reconciles, finally on the fifth reconcile and a total of 45 seconds, the COSI user is created. This timing is very consistent in my minikube environment every time I create a test object store.

Here are the failure logs that cause the reconcile to restart, then finally succeed. See attached full [operator.log](https://github.com/rook/rook/files/14544143/operator.log)
```
2024-03-08 23:34:27.907685 E | ceph-object-controller: failed to reconcile CephObjectStore "rook-ceph/my-store". failed to create object store deployments: failed to get COSI user "cosi": Get "http://rook-ceph-rgw-my-store.rook-ceph.svc:80/admin/user?format=json&uid=cosi": dial tcp 10.108.29.221:80: connect: connection refused
2024-03-08 23:34:40.336048 E | ceph-object-controller: failed to reconcile CephObjectStore "rook-ceph/my-store". failed to create object store deployments: failed to get COSI user "cosi": Get "http://rook-ceph-rgw-my-store.rook-ceph.svc:80/admin/user?format=json&uid=cosi": dial tcp 10.108.29.221:80: connect: connection refused
2024-03-08 23:34:52.416211 E | ceph-object-controller: failed to reconcile CephObjectStore "rook-ceph/my-store". failed to create object store deployments: failed to get COSI user "cosi": Get "http://rook-ceph-rgw-my-store.rook-ceph.svc:80/admin/user?format=json&uid=cosi": dial tcp 10.108.29.221:80: connect: connection refused
2024-03-08 23:35:03.410629 E | ceph-object-controller: failed to reconcile CephObjectStore "rook-ceph/my-store". failed to create object store deployments: failed to get COSI user "cosi": Get "http://rook-ceph-rgw-my-store.rook-ceph.svc:80/admin/user?format=json&uid=cosi": dial tcp 10.108.29.221:80: connect: connection refused
2024-03-08 23:35:14.378067 I | ceph-object-controller: creating COSI user "cosi"
```

**Expected behavior:**
In the common case, the COSI user should be created successfully without causing so many retries. If there is a known reason the user cannot be created for some time after the object store is created, let's add a check for that condition in the reconcile instead of filling the logs with failed reconciles.

**How to reproduce it (minimal and precise):**

1. Create a ceph cluster (cluster-test.yaml)
2. Create an object store (object-test.yaml)
3. Check the operator logs for the creation of the COSI user

**Comment by thotz** [comment]
Even rgw pod is up and running it may take few seconds to be ready, hence this request is failing IMO

**Comment by travisn** [comment]
> Even rgw pod is up and running it may take few seconds to be ready, hence this request is failing IMO. May be something like this need to be added before the creating the cosi user

45s is a "long" time to wait since the rgw pod is already up. We need to understand why so long and if anything can be improved with this. In the normal install flow we want to avoid reconcile failures if possible. I believe other users can be created immediately after that without waiting this long, but will check for sure...

**Comment by huv95** [comment]
I am running ceph object store with hdd drives, named s3. 
To improve speed, I just added some ssd drives to the cluster, and created a new object store called s3-ssd.
But, it is not created successfully, I get the error "failed to create object store deployments: `failed to create COSI user "cosi"`. 
Do you have any advice?

**Comment by travisn** [comment]
@huv95 Does the object store reconcile never succeed? Or you see that message continuously, and even after restarting the operator pod? If it is continuous, there must be some other error configuring the object store. For example, do you have at least three OSDs on different nodes? What does `ceph status` show? Can you share more of the operator log?

**Comment by kalmufti** [comment]
I'm having this issue, been trying to get multisite store to work by following the guides in documentations. Here is some logs: 

```bash
bash-4.4$ ceph status
  cluster:
    id:     ff43cb63-97f8-4f2f-bad2-6cf8d3cc074f
    health: HEALTH_OK

  services:
    mon: 3 daemons, quorum a,b,c (age 16h)
    mgr: a(active, since 42m), standbys: b
    osd: 9 osds: 9 up (since 16h), 9 in (since 16h)
    rgw: 1 daemon active (1 hosts, 1 zones)

  data:
    pools:   10 pools, 289 pgs
    objects: 253 objects, 19 MiB
    usage:   609 MiB used, 9.2 TiB / 9.2 TiB avail
    pgs:     289 active+clean
```

```bash
2024-07-16 23:49:24.654448 I | ceph-object-controller: created object store "zone-b-multisite-store" in namespace "rook-ceph"
2024-07-16 23:49:25.506565 I | ceph-object-controller: creating COSI user "cosi"
2024-07-16 23:49:26.561448 E | ceph-object-controller: failed to reconcile CephObjectStore "rook-ceph/zone-b-multisite-store". failed to create object store deployments: failed to create COSI user "cosi": UnknownError tx0000014363ecb19d05e0f-0066970705-1b081-zone-b 1b081-zone-b-zonegroup-a
```
Same error if I try to create the user from the dashboard:

```bash
RGW REST API failed request with status code 500 (b'{"Code":"UnknownError","Message":"","RequestId":"tx000004eb3e429cdc2c977-006' b'6970620-1b081-zone-b","HostId":"1b081-zone-b-zonegroup-a"}')
```

I was able to create local buckets on my two clusters, but getting multisite to work has been a pain.

**Comment by travisn** [comment]
@thotz Could you take a look at this failure for multisite?

**Comment by thotz** [comment]
The ideally user should created by master zone, we have check for whether user exists then only it will try to create the user IMO, @kalmufti can give more details about ur configuration. How multisite is configured is it two different clusters, or same cluster with different namespaces or same namespace in same cluster

**Comment by kalmufti** [comment]
@thotz, I'm trying to run multisite on different clusters, using the same namespace ```rook-ceph``` on both. The clusters uses Cilium as CNI with IPAM LoadBalancer, both clusters runs unique networks for both pods and services, and all communications between the clusters happens over BGP.

Here are the steps I'm following to setup multisite using the official docs:
1. Deploy rook operator, crds, cluster, etc, on both kube clusters.
2. Deploy the ```object-multisite.yaml``` example file on the first kube cluster.
3. Create an IPAM address for the multisite ```multisite-store```.
4. Test the ip address with ```curl``` to confirm external access to the store.
5. Extract ```realm-a-keys```.
6. Update the example ```object-multisite-pull-realm.yaml``` with the ip address url and the realm keys.
7. Update the example ```object-multisite-pull-realm.yaml``` namespace from ```rook-ceph-secondary``` to ```rook-ceph``` to match the default namespace for the second cluster.
8. Deploy ```object-multisite-pull-realm.yaml``` to the second cluster.

Sometimes I had to kill the operator on the pulling site to get to pass another error where to complains that ```CephObjectZone - zonegroup-a``` dosen't exist.

I've spent over a week trying to get multisite to work, I've tried everything comes to mind.

**Comment by guchengod** [comment]
I encountered the same problem, but the problem was solved after I deleted the Secret named "rook-ceph-object-user-my-store-cosi"

**Comment by BlaineEXE** [comment]
Will be resolved by #15035

**Comment by travisn** [comment]
#15035 was merged
